File: 230_KthSmallestElementInABST/c++/KthSmallestElement.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int kthSmallest(TreeNode* root, int k) {
        int ans = 0;
        bool res = recruciveTravel(root, k, ans);
        return ans;
    }
private:
    //LDR
    bool recruciveTravel(TreeNode* node, int& k, int& ans){
        //left sub-tree
        if(node->left!=NULL){
            if(recruciveTravel(node->left, k, ans)){
                return true;
            }
        }
        
           
        if(k==1){
            ans = node->val;
            return true;
        }
        
        k--;
         
        //right sub-tree
        if(node->right!=NULL){
            if(recruciveTravel(node->right, k, ans)){
                return true;
            }
        }
        return false;
           
    }
};

File: 230_KthSmallestElementInABST/c#/KthSmallestElement.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public int KthSmallest(TreeNode root, int k) {
        int[] ans = {k,0};
        recrusiveTravel(root, ans);
        return ans[1];
    }
    
    public bool recrusiveTravel(TreeNode node, int[] k){
        if(node.left!=null){
            if(recrusiveTravel(node.left, k)){
                return true;
            }
        }
        
        if(k[0] == 1){
            k[1] = node.val;
            return true;
        }
        
        k[0]--;
        
        if(node.right!=null){
            if(recrusiveTravel(node.right, k)){
                return true;
            }
        }
        
        return false;
    }
}